File: experiments/runs/2026-09-16_16-45-11_claim-office-example-mapping_exact-sol-v1.6-test-list-dimensions-cc_opus-5-no-thinking/src/reimbursement.ts
import { type Item } from "./price-list.js";

const HIGH_ENCHANTMENT_CLAIM_THRESHOLD = 8;
const HIGH_ENCHANTMENT_REIMBURSEMENT_RATE = 0.5;
const FULL_REIMBURSEMENT_RATE = 1;

function isHighlyEnchanted(item: Item): boolean {
  return (item.enchantment ?? 0) >= HIGH_ENCHANTMENT_CLAIM_THRESHOLD;
}

/**
 * The share of a damage the MHPCO reimburses.
 *
 * Two special clauses exist. Damage to an item with enchantment level >= 8 is
 * reimbursed at 50 %. Damage to an item of dragon material is fully reimbursed
 * -- which is also the rate for an item under no special clause, so the clause
 * changes an outcome only where the high-enchantment clause would otherwise
 * reduce it, and there the high-enchantment clause wins. The dragon-material
 * clause therefore needs no branch of its own.
 */
function reimbursementRateFor(item: Item): number {
  return isHighlyEnchanted(item)
    ? HIGH_ENCHANTMENT_REIMBURSEMENT_RATE
    : FULL_REIMBURSEMENT_RATE;
}

/** The reimbursable share of a damage amount, before the deductible. */
export function reimbursableAmount(damageAmount: number, item: Item): number {
  return damageAmount * reimbursementRateFor(item);
}

File: experiments/runs/2026-09-20_07-23-29_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/Reimbursement.java
/**
 * The MHPCO's reimbursement clauses: what share of a damage it pays for an
 * item, before the deductible. Damage to a severely enchanted item is
 * reimbursed at half; dragon material is fully reimbursed, as is any item no
 * clause singles out. Where both clauses apply, the severe-enchantment clause
 * wins.
 */
public final class Reimbursement {

    private static final double FULL = 1.0;
    private static final double SEVERE_ENCHANTMENT_SHARE = 0.50;
    private static final int SEVERE_ENCHANTMENT_THRESHOLD = 8;

    private Reimbursement() {
    }

    public static double shareFor(Item item) {
        if (isSeverelyEnchanted(item)) {
            return SEVERE_ENCHANTMENT_SHARE;
        }
        return FULL;
    }

    private static boolean isSeverelyEnchanted(Item item) {
        return item.enchantment() >= SEVERE_ENCHANTMENT_THRESHOLD;
    }
}
